src/kitovu/sync/kitovu_filecache.py

```diff
@@ -0,0 +1,113 @@
+            return Filestate.NONE
+        elif (remote_digest != cached_digest) and (local_digest == cached_digest):  # case 4
+            return Filestate.REMOTE
+        elif (remote_digest == cached_digest) and (local_digest != cached_digest):  # case 6
+            return Filestate.LOCAL
+        elif (remote_digest != cached_digest) and (local_digest != cached_digest):  # case 7
+            return Filestate.BOTH
+
+        # new file B is discovered when remote_digest != cached_digest
+
+        # Cases:
+        # 1. remote deleted (triggers exception), local exists => REMOTE*
+        # 2. remote deleted (triggers exception), local exists AND changed (local_digest and cached_digest differ) => BOTH*
+        # case 1 and 2 special, dealt with separately, FIXME externally
+
+        # Normal Cases:
+        # 3. new file remote, local none => REMOTE, download FIXME externally
+        # 4. remote B, local A => remote_digest and local digest differ, but local digest and cached digest same => REMOTE, download
+
+        # 5. remote and local same  => NONE
+        # 6. remote A unchanged, local changed => remote_digest and cached_digest same, but local_digest and cached_digest differ => LOCAL
+        # 7. remote B, local changed => remote_digest and cached_digest differ, local_digest and cached_digest differ => BOTH
+        # 8. remote B, local b => remote_digest and cached_digest same, but local_digest and cached_digest differ => case 6, update Filecache
+        # case 8 is basically the same as 6, it simply triggers the user's decision and needs to update the filecache
+
+    def write(self) -> None:
+        """"Writes the data-dict to JSON."""
+        json_data: typing.Dict[str, typing.Dict[str, str]] = {}
+
+        for key, value in self._data.items():
+            json_data[str(key)] = value.to_dict()
+
+        with self._filename.open("w") as f:
+            json.dump(json_data, f)
+
+    def load(self) -> None:
+        """This is called first when the synchronisation process is started."""
+        with self._filename.open("r") as f:
+            json_data = json.load(f)
+        for key, value in json_data.items():
+            digest: str = value["digest"]
+            plugin: str = value["plugin"]
+            self._data[pathlib.Path(key)] = File(local_digest_at_synctime=digest, plugin=plugin)
+            # FIXME should be of type AbstractSyncPlugin - conversion needed?
+
+    def update(self, path: pathlib.Path, plugin: syncplugin.AbstractSyncPlugin, local_digest_at_synctime):
+        file = File(local_digest_at_synctime=local_digest_at_synctime, plugin=plugin)
+        self._data.update(path, file)
+
+    def discover_changes(self, path: pathlib.Path, plugin: syncplugin.AbstractSyncPlugin) -> Filestate:
+        """checks if the file that is currently downloaded (path-argument) has changed
+        in comparison to the local filecache and local file)."""
```

Comment:
The typical format for docstrings is similar to a git commit (or email with subject/body) - a one-line summary, blank line, and then a longer explanation, all in imperative form:

```
"""Check if the file currently being downloaded has changed.

(Longer text here)
"""
```

Also see [PEP 257](https://www.python.org/dev/peps/pep-0257/) which is basically PEP8 but for docstrings.